ci/publish.rs

```diff
@@ -0,0 +1,242 @@
+                        .replace("\"", "")
+                        .trim()
+                        .to_string(),
+                );
+            }
+        }
+        let name = name.unwrap();
+        let version = version.unwrap();
+        let next_version = if CRATES_TO_PUBLISH.contains(&&name[..]) {
+            bump(&version)
+        } else {
+            version.clone()
+        };
+        Crate {
+            manifest: manifest.to_path_buf(),
+            name,
+            version,
+            next_version,
+        }
+    }
+}
+
+fn bump_version(krate: &Crate, crates: &[Crate]) {
+    let contents = fs::read_to_string(&krate.manifest).unwrap();
+
+    let mut new_manifest = String::new();
+    let mut is_deps = false;
+    for line in contents.lines() {
+        let mut rewritten = false;
+        if line.starts_with("version =") {
+            if CRATES_TO_PUBLISH.contains(&&krate.name[..]) {
+                println!(
+                    "bump `{}` {} => {}",
+                    krate.name, krate.version, krate.next_version
+                );
+                new_manifest.push_str(&line.replace(&krate.version, &krate.next_version));
+                rewritten = true;
+            }
+        }
+
+        is_deps = if line.starts_with("[") {
+            line.contains("dependencies")
+        } else {
+            is_deps
+        };
+
+        for other in crates {
+            if !is_deps || !line.starts_with(&format!("{} ", other.name)) {
+                continue;
+            }
+            if !line.contains(&other.version) {
+                if !line.contains("version =") {
+                    continue;
+                }
+                panic!(
+                    "{:?} has a dep on {} but doesn't list version {}",
+                    krate.manifest, other.name, other.version
+                );
+            }
+            rewritten = true;
+            new_manifest.push_str(&line.replace(&other.version, &other.next_version));
+            break;
+        }
+        if !rewritten {
+            new_manifest.push_str(line);
+        }
+        new_manifest.push_str("\n");
+    }
+    fs::write(&krate.manifest, new_manifest).unwrap();
+}
+
+/// Bumps a semver version to the next semver version that we should publish.
+///
+/// Currently this bumps the minor number, which for cranelift changes X.Y.Z to
+/// X.Y+1.Z, e.g. 0.49.0 to 0.50.0
+fn bump(version: &str) -> String {
+    let mut iter = version.split('.').map(|s| s.parse::<u32>().unwrap());
+    let major = iter.next().expect("major version");
+    let minor = iter.next().expect("minor version");
+    format!("{}.{}.0", major, minor + 1)
+}
+
+fn publish(krate: &Crate, dry_run: bool, any_failed: &mut bool) {
+    if !CRATES_TO_PUBLISH.iter().any(|s| *s == krate.name) {
+        return;
+    }
+
+    // Skip the simplejit crate since it tries to generate a lock file and may
+    // point to unpublished dependencies. Only skip it in dry run mode though.
```

Comment:
To my knowledge, cranelift-simplejit doesn't do anything special wrt cargo, and doesn't have any unusual dependencies. Can you point to what it is about this crate that causes problems?

Reply:
It's sort of a weird part of Cargo, but if Cargo for some crates decides to put a `Cargo.lock` in the package, meaning it needs to resolve the crate graph. If a crate depends on something not published on crates.io, however, as would be the case in the "dry run" mode which happens on PRs, then this step will fail because `Cargo.lock` cannot be generated. I hit this locally when I prefixed all the crates with `alexcrichton-*` and nothing existed on crates.io.

Cargo decides to put `Cargo.lock` in crates that contains examples or binaries, and simplejit happened to be the only cranelift crate that met that criteria (had an example)